### PR TITLE
feat: Add support for C++ TurboModules Autolinking

### PR DESCRIPTION
**Pull request by cortinico**
Summary:
---------

This adds support for C++ TurboModules Autolinking on Android.

I added those properties in the config command:

#### platforms.android.cxxModuleCMakeListsModuleName

> Note: Only applicable when new architecture is turned on and for C++ TurboModules.

The name of the CMake target that is building the C++ Turbo Module. This is generally
different from the codegen target which is called `react_codegen_<yourlibrary>`.
You will need to create a custom `CMakeLists.txt`, which defines a target, depends on `react_codegen_<yourlibrary>` and builds your C++ Turbo Module.
Specify the name of the target here so that it can be properly linked to the rest of the app project.

#### platforms.android.cxxModuleCMakeListsPath

> Note: Only applicable when new architecture is turned on and for C++ TurboModules.

A relative path to a custom _CMakeLists.txt_ file that is used to build a C++ TurboModule. Relative to `sourceDir`.

#### platforms.android.cxxModuleHeaderName

> Note: Only applicable when new architecture is turned on and for C++ TurboModules.

The name of the C++ TurboModule. Your C++ TurboModule generally should implement one of the JSI interface generated by Codegen. You need to specify the name of the class that is implementing the generated interface here, so that it can properly be linked in the rest of the app project.

We expect that your module exposes a filed called `kModuleName` which is a string that contains the name of the module. This field is properly populated by codegen, but if you don't use codegen, you will have to make sure that this field is present.

Say your module name is `NativeCxxModuleExample`, it will be included in the autolinking logic as follows:

```cpp
// We will include a header called as your module
#include <NativeCxxModuleExample.h>

...

std::shared_ptr<TurboModule> rncli_cxxModuleProvider(
    const std::string& name,
    const std::shared_ptr<CallInvoker>& jsInvoker) {

  // Here we expect you have a field called kModuleName
  if (name == NativeCxxModuleExample::kModuleName) {
    // Here we initialize your module
    return std::make_shared<NativeCxxModuleExample>(jsInvoker);
  }
  return nullptr;
}
```

Test Plan:
----------

I had to test it manually in a nightly project by tampering the `node_modules` folder and I managed to make it work. Not sure how I can test it more.

Checklist
----------

- [x] Documentation is up to date to reflect these changes.
- [x] Follows commit message convention described in [CONTRIBUTING.md](https://github.com/react-native-community/cli/blob/main/CONTRIBUTING.md#commit-message-convention)

### CODE REVIEW

**Comment by thymikee** [comment]
So happy to see this! Let me know if we can help anyhow

**Comment by cortinico** [comment]
This works in conjusction with:
- https://github.com/facebook/react-native/pull/43049

File: packages/cli-platform-android/native_modules.gradle [hunk 1]
```diff
@@ -259,6 +266,7 @@ class ReactNativeModules {
     }
   }
 
+
```
Comment (szymonrybczak) [inline packages/cli-platform-android/native_modules.gradle:269]: nit: useless new line

File: packages/cli-platform-android/native_modules.gradle [hunk 2]
```diff
@@ -398,7 +432,7 @@ class ReactNativeModules {
     String[] nodeCommand = ["node", "-e", cliResolveScript]
     def cliPath = this.getCommandOutput(nodeCommand, this.root)
 
-    String[] reactNativeConfigCommand = ["node", cliPath, "config"]
+    String[] reactNativeConfigCommand = ["cat", "/Users/ncor/config.json"]
```
Comment (szymonrybczak) [inline packages/cli-platform-android/native_modules.gradle:435]: local testing leftoever I guess :D

Comment (cortinico) [reply packages/cli-platform-android/native_modules.gradle]: fixed

File: packages/cli-platform-android/native_modules.gradle (hunk 1 above)
Comment (cortinico) [reply packages/cli-platform-android/native_modules.gradle:268]: fixed

File: docs/dependencies.md [hunk 3]
```diff
@@ -140,3 +143,47 @@ An array of custom component descriptor strings. By default they're generated ba
...
+
+A relative path to a custom _CMakeLists.txt_ file that is used to build a C++ TurboModule. Relative to `sourceDir`.
+
+#### platforms.android.cxxModuleHeaderName
+
+> Note: Only applicable when new architecture is turned on and for C++ TurboModules.
+
+The name of the C++ TurboModule. Your C++ TurboModule generally should implement one of the JSI interface generated by Codegen. You need to specify the name of the class that is implementing the generated interface here, so that it can properly be linked in the rest of the app project.
+
+We expect that your module exposes a filed called `kModuleName` which is a string that contains the name of the module. This field is properly populated by codegen, but if you don't use codegen, you will have to make sure that this field is present.
```
Comment (szymonrybczak) [inline docs/dependencies.md:168]:
```suggestion
We expect that your module exposes a field called `kModuleName` which is a string that contains the name of the module. This field is properly populated by codegen, but if you don't use codegen, you will have to make sure that this field is present.
```

**Comment by WoLewicki** [comment]
@cortinico did you maybe check what was needed for the autolinking to work in this PR: https://github.com/software-mansion/react-native-gesture-handler/pull/2708? Would be nice to cover all the needed parts 🎉

**Comment by cortinico** [comment]
> @cortinico did you maybe check what was needed for the autolinking to work in this PR: [software-mansion/react-native-gesture-handler#2708](https://github.com/software-mansion/react-native-gesture-handler/pull/2708)? Would be nice to cover all the needed parts 🎉

@WoLewicki I've briefly checked and it seems like Gesture Handler would fit the current setup for TM CXX Autolinking I'm proposing here (we'll have to adapt the setup a bit)

**Review by szymonrybczak — APPROVED** [review]
:shipit:

**Comment by mrousavy** [comment]
Is there any example C++ TurboModule that autolinks? I can't seem to get the CMakeLists configuration right for react-native-mmkv...

**Comment by cortinico** [comment]
> Is there any example C++ TurboModule that autolinks? I can't seem to get the CMakeLists configuration right for react-native-mmkv...

Sadly not. I had it working on my local setup, and I haven't had the time to work on samples or further docs.

**Comment by mrousavy** [comment]
Is my understanding correct that there are 3 CMakeLists.txt files per CxxTurboModule?

1. One in `cpp/CMakeLists.txt` that builds the TurboModule (incl. all dependencies like MMKVCore, Skia, OpenGL or whatever)
2. One in `android/CMakeLists.txt` that just depends on the one in `../cpp/`
3. One generated by CodeGen (`android/build/generated/...`)

And then my react-native.config.js should look like this:

```js
// https://github.com/react-native-community/cli/blob/main/docs/dependencies.md

module.exports = {
  dependency: {
    platforms: {
      /**
       * @type {import('@react-native-community/cli-types').IOSDependencyParams}
       */
      ios: {},
      /**
       * @type {import('@react-native-community/cli-types').AndroidDependencyParams}
       */
      android: {
        cxxModuleCMakeListsModuleName: 'reactnativemmkv',
        cxxModuleCMakeListsPath: './CMakeLists.txt',
        cxxModuleHeaderName: 'NativeMmkvModule',
      },
    },
  },
};
```

Not sure what I'm missing, but I can't seem to get that to build/link properly... Code is here: [react-native-mmkv (`feat/turbomodule`)](https://github.com/mrousavy/react-native-mmkv/tree/feat/turbomodule)

**Pull request merged**